### DISCUSSION TITLE
Callback flow in slint file

### DISCUSSION THREAD

**Discussion by qarmin** [slint-ui/slint#3947] [Q&A]
When using callbacks I found that sometimes it is hard to use them and I need to use workarounds
- it is possible to connect two callbacks?
```
component c1 {
     callback something_changed();
}
component c2 {
    callback something_changed();
    c1 {
         something_changed <=> root.something_changed(); // Or => <=(one way callback direction)
    }
}
```
is this possible?
Now I need to use (with more variables in callback this is more problematic)
```
    c1 {
         something_changed() => {
               root.something_changed()
         }
    }
```
- It is possible to connect to global callback?
```
component c1 {
     Global.callback() => {
           debug!("Executed global callback");
     }
}
```

**Comment by bjorn** [comment]
> it is possible to connect two callbacks?

You can create a [callback alias](https://slint.dev/releases/1.3.0/docs/slint/src/language/syntax/callbacks#aliases), but it needs to be done the other way around and without parenthesis:

```
component c1 {
    callback something_changed();
}
component c2 {
    callback something_changed <=> instance1.something_changed;
    instance1 := c1 {}
}
```

> * It is possible to connect to global callback?

I'm still a bit new to Slint so I may be wrong, but I believe a callback is not to be confused with a signal. So you can't "connect" to a callback, you can only set one handler for it, and it needs to be done at the instantiation. Since a global gets instantiated implicitly, I think you can only set the handler for its callbacks through the API.

I've tried to re-export a global callback [in SlintPad](https://slintpad.com/?snippet=import+%7B+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Aglobal+Logic++%7B%0A++++callback+magic-operation%3B%0A%7D%0A%0Aexport+component+Foo+%7B%0A++++callback+magic-operation+%3C%3D%3E+Logic.magic-operation%3B%0A%7D%0A%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++text-item+%3A%3D+Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++%7D%0A++++++++Foo+%7B%0A++++++++++++magic-operation+%3D%3E+%7B%0A++++++++++++++++text-item.text+%3D+%22Foo%22%3B%0A++++++++++++%7D%0A++++++++%7D%0A++++++++Button+%7B%0A++++++++++++clicked+%3D%3E+%7B+Logic.magic-operation%28%29%3B+%7D%0A++++++++++++text%3A+%22Click+me%22%3B%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=native) to see if this would allow defining a handler in Slint, but while the example compiled, the callback handler does not appear to work. I wonder if this is a bug or whether there is a reason why that can't work, or whether I made a mistake of course.

(edit: well I guess the reason is kind of obvious... if that would actually work, it would allow possibly defining multiple handlers and there would be no clear way of choosing which one should be used)

(edit2: though... the same is actually true for callback aliases in general, and when [creating multiple aliases to a regular callback](https://slintpad.com/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0A%0Aexport+component+Foo+%7B%0A++++callback+clickedA+%3C%3D%3E+button.clicked%3B%0A++++callback+clickedB+%3C%3D%3E+button.clicked%3B%0A++++button+%3A%3D+Button+%7B%0A++++++++text%3A+%22OK%21%22%3B%0A++++%7D%0A%7D%0A%0Aexport+component+Demo+%7B%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++text1+%3A%3D+Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++%7D%0A++++++++text2+%3A%3D+Text+%7B%0A++++++++++++text%3A+%22Hello+World%21%22%3B%0A++++++++++++font-size%3A+24px%3B%0A++++++++++++horizontal-alignment%3A+center%3B%0A++++++++%7D%0A++++++++AboutSlint+%7B%0A++++++++++++preferred-height%3A+150px%3B%0A++++++++%7D%0A++++++++HorizontalLayout+%7B+alignment%3A+center%3B+foo+%3A%3D+Foo+%7B%0A++++++++++++clickedA+%3D%3E+%7B+text1.text+%3D+%22A%22%3B+%7D%0A++++++++++++clickedB+%3D%3E+%7B+text2.text+%3D+%22B%22%3B+%7D%0A++++++++%7D%7D%0A++++%7D%0A%7D%0A&style=native) it just seems to be kind of arbitrary, which one will actually fire)

**Reply by ogoffart** [reply]
This is exposing some bugs.
I think I should at least implement something so it throws a warning when aliasing to such callback.

**Reply by bjorn** [reply]
Hmm, well aliasing a global callback is explicitly mentioned as a feature in the documentation, for the purpose of exposing a callback from a non-exported global. But I guess this currently only works when the handler is set from the application rather than in Slint (I haven't tried).

**Reply by ogoffart** [reply]
I filled an issue to track that in https://github.com/slint-ui/slint/issues/3966
